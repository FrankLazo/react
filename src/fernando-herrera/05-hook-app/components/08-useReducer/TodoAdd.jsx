import useForm from '../../hooks/useForm';

const TodoAdd = ({ handleAddTodo }) =>
{
	const { values: { description }, handleInputChange, reset } = useForm({ description: '' });

	const handleSubmit = (e) =>
	{
		e.preventDefault();

		if (description.trim().length <= 1)
		{
			return;
		}

		const newTodo = {
			id:          new Date().getTime(),
			description: description,
			done:        false,
		};

		handleAddTodo(newTodo);
		reset();
	};

	return (
		<form onSubmit={ handleSubmit } className="form">
			<div className="field">
				<div className="control">
					<input
						type="text"
						name="description"
						placeholder="Learn..."
						autoComplete="off"
						value={ description }
						onChange={ handleInputChange }
						className="input"
						/>
				</div>
			</div>

			<div className="field">
				<div className="control">
					<button
						type="submit"
						className="button is-primary"
						>
						Add task
					</button>
				</div>
			</div>
		</form>
	);
};

export default TodoAdd;
