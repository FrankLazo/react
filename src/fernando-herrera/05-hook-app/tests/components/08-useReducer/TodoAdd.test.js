import { shallow } from 'enzyme';
import TodoAdd from '../../../components/08-useReducer/TodoAdd';

describe('<TodoAdd /> tests', () => {
	const handleAddTodo = jest.fn();
	const wrapper = shallow(<TodoAdd handleAddTodo={ handleAddTodo } />);

	test('should show the right component', () => {
		expect(wrapper).toMatchSnapshot();
	});

	test('should not call handleAddTodo', () => {
		const formSubmit = wrapper.find('form').prop('onSubmit');
		formSubmit( { preventDefault(){} } );

		expect(handleAddTodo).not.toHaveBeenCalled();
	});

	test('should call handleAddTodo', () => {
		const value = 'Learn React';
		const input = wrapper.find('input');

		input.simulate('change', { target: { value, name: 'description' } });

		const formSubmit = wrapper.find('form').prop('onSubmit');
		formSubmit( { preventDefault(){} } );

		expect(handleAddTodo).toHaveBeenCalledTimes(1);
		expect(handleAddTodo).toHaveBeenCalledWith( expect.any(Object) );
		expect(handleAddTodo).toHaveBeenCalledWith({
			id: expect.any(Number),
			description: value,
			done: false,
		});
		expect(input.prop('value')).toBe('');	// Para verificar el efecto de reset()
	});
});
