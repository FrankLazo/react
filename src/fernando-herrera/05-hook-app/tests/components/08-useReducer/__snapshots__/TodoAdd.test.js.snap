// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoAdd /> tests should show the right component 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <div
    className="field"
  >
    <div
      className="control"
    >
      <input
        autoComplete="off"
        className="input"
        name="description"
        onChange={[Function]}
        placeholder="Learn..."
        type="text"
        value=""
      />
    </div>
  </div>
  <div
    className="field"
  >
    <div
      className="control"
    >
      <button
        className="button is-primary"
        type="submit"
      >
        Add task
      </button>
    </div>
  </div>
</form>
`;
